src/Compilers/Core/MSBuildTask/GenerateMSBuildAnalyzerConfig.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    /// <summary>
+    /// Transforms a set of MSBuild Properties and Metadata into a global analyzer config.
+    /// </summary>
+    /// <remarks>
+    /// This task takes a set of items passed in via <see cref="MetadataItems"/> and <see cref="PropertyItems"/> and transforms
+    /// them into a global analyzer config. 
+    /// 
+    /// <see cref="PropertyItems"/> is expected to be a list of items whose <see cref="ITaskItem.ItemSpec"/> is the property name
+    /// and have a metadata value called <c>Value</c> that contains the evaluated value of the property. Each of the ]
+    /// <see cref="PropertyItems"/> will be transformed into an <c>msbuild_property.<em>ItemSpec</em> = <em>Value</em></c> entry in the
+    /// global section of the generated config file.
+    /// 
+    /// <see cref="MetadataItems"/> is expected to be a list of items whose <see cref="ITaskItem.ItemSpec"/> represents a file in the 
+    /// compilation source tree. It should have two metadata values: <c>ItemType</c> is the name of the MSBuild item that originally 
+    /// inlcuded the file (e.g. <c>Compile</c>, <c>AdditionalFile</c> etc.); <c>MetadataName</c> is expected to contain the name of
+    /// another piece of metadata that should be retreived and used as the output value in the file. It is expected that a given 
+    /// file can have multiple entries in the <see cref="MetadataItems" /> differing by its <c>ItemType</c>.
+    /// 
+    /// Each of the <see cref="MetadataItems"/> will be transformed into a new section in the generated config file. The section
+    /// header will be the full path of the item (generated via its<see cref="ITaskItem.ItemSpec"/>), and each section will have a 
+    /// set of <c>msbuild_metadata.<em>ItemType</em>.<em>MetadataName</em> = <em>RetreivedMetadataValue</em></c>, one per <c>ItemType</c>
+    /// 
+    /// The Microsoft.Managed.Core.targets calls this task with the collected results of the <c>AnalyzerProperty</c> and 
+    /// <c>AnalyzerItemMetadata</c> item groups. 
+    /// </remarks>
+    public sealed class GenerateMSBuildAnalyzerConfig : Task
```

Comment:
One part I'm unclear on is what is the responsibility of this type with respect to generating a well formed / correct editorconfig file? Is the expectation that this `Task` is best effort, guaranteed to produce correct editor config, etc ...?

Reply:
It *should* be a valid config, although there is the newline bug that means the files it produces _can_ be incorrect. Correct parsing of editorconfig is supposed to just ignore lines it can't understand though, so even then, it'll be valid you just can't read the whole property.

Want me to add something to the task description that explains it?

---
In reply to: [419847995](https://github.com/dotnet/roslyn/pull/43617#discussion_r419847995) [](ancestors = 419847995)

Reply:
Possibly add a remarks section essentially noting that there will always be cases where we can produce invalid editorconfig files. We can definitely get the 99% case covered but I'm sure there are ways that developers could do evil things to produce an invalid ones. 

Another question, @jasonmalinowski brought this up in chat. When the editor config is invalid how will the customer find out? Will there just be an error message in the diagonstics for example?

Reply:
An editorconfig parsed is supposed to ignore any lines it doesn't understand when parsing: I've checked and we can handle a malformed config; we just ignore anything we can't parse. Of course that isn't particularly helpful for the user to understand when things aren't working. We 'blessed' the diagnostic options as having warnings when we don't understand them, and I expect we'll do the same for the msbuild properties generated here. The task should always generate *something* we can recognize (`msbuild_property` / `msbuild_metadata`) so when it comes to parsing those (another PR) will probably issue warnings for things we can't understand.

---
In reply to: [420302735](https://github.com/dotnet/roslyn/pull/43617#discussion_r420302735) [](ancestors = 420302735,419847995)

Reply:
An editorconfig parsed is supposed to ignore any lines it doesn't understand when parsing: I've checked and we can handle a malformed config; we just ignore anything we can't parse. Of course that isn't particularly helpful for the user to understand when things aren't working. We 'blessed' the diagnostic options as having warnings when we don't understand them, and I expect we'll do the same for the msbuild properties generated here. The task should always generate *something* we can recognize (`msbuild_property` / `msbuild_metadata`) so when it comes to parsing those (another PR) will probably issue warnings for things we can't understand.

---
In reply to: [420351311](https://github.com/dotnet/roslyn/pull/43617#discussion_r420351311) [](ancestors = 420351311)